src/dhtproto/client/internal/NodeHashRanges.d

```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+            node_hash_ranges = buffer to receive hash range information of
+                nodes which cover the specified hash
+            use_node = delegate to call to gain access to a request-on-conn to
+                communicate with the selected node
+            get = delegate to communicate with the selected node (over the
+                request-on-conn provided by use_node). Should return true if the
+                get succeeded or false to try the next node
+
+        Returns:
+            true if there are no nodes covering the specified hash
+
+    ***************************************************************************/
+
+    public bool getFromNode ( hash_t h, ref NodeHashRange[] node_hash_ranges,
+        UseNodeDg use_node,
+        bool delegate ( RequestOnConn.EventDispatcher ) get )
+    {
+        auto nodes = this.getNodesForHash(h, node_hash_ranges);
+        if ( nodes.length == 0 )
+            return true;
+
+        foreach_reverse ( node_hash_range; nodes )
+        {
+            bool try_next_node;
+            use_node(node_hash_range.addr,
+                ( RequestOnConn.EventDispatcher conn )
+                {
+                    try_next_node = get(conn);
+                }
+            );
+
+            // If we got the record or an error occurred, don't try more nodes
+            if ( !try_next_node )
+                break;
+        }
+
+        return false;
+    }
+
+    /***************************************************************************
+
+        Helper encapsulating the node-selection logic required by requests that
+        put a single record to the DHT. Namely:
+
+            During a data redistribution, more than one node may be responsible
+            for a given key. In this case, the record is written to the node
+            that was most recently reported as being responsible for the key.
+
+        Params:
+            h = hash to query
+            node_hash_ranges = buffer to receive hash range information of
+                nodes which cover the specified hash
+            use_node = delegate to call to gain access to a request-on-conn to
+                communicate with the selected node
+            put = delegate to communicate with the selected node (over the
+                request-on-conn provided by use_node)
+
+        Returns:
+            true if there are no nodes covering the specified hash
+
+    ***************************************************************************/
+
+    public bool putToNode ( hash_t h, ref NodeHashRange[] node_hash_ranges,
+        UseNodeDg use_node,
+        void delegate ( RequestOnConn.EventDispatcher ) put )
+    {
+        auto nodes = this.getNodesForHash(h, node_hash_ranges);
+        if ( nodes.length == 0 )
+            return true;
+
+        size_t newest;
+        ulong newest_order;
+        foreach ( i, node_hash_range; nodes )
+        {
+            if ( node_hash_range.order > newest_order )
```

Comment:
To confirm: `node_hash_range.order` starts from `1`? Maybe it's safer to do `if (node_hash_range.order >= newest_order)`.

Reply:
There should never be two nodes with the same `order`. I can add checks that verify this.

Reply:
No, it was my brain that stuck yesterday. This piece of code is actually fine.